### PR TITLE
implementation of SPD manifolds

### PR DESCRIPTION
**Pull request by tao-harald**
I'm trying to implement the SPD manifolds, which is widely used in many papers. 

Current commits have implemented all abstract methods of `base.Manifold`, while more method implementations are on the way.
Some test has been done locally and constructions of the testing module are also on the way.

### Progress
 * [x] move symmetric matrix operations to `batch_linalg`
 * [x] `keepdims` functionality  for `inner` and `_stein_metric`.
 * [x] mention in documentation about SPD manifolds.
 * [x] implementation for `random` and `origin`
 * [x] manifold test module.
    * [x] shape case for a symmetric positive-definite matrix.
    * [x] test basic operation of SPD manifolds.
    * [x] simple optimization problem on SPD manifolds.
 * [ ] mention the PR in the `CHANGELOG`.

### Reference Implementation
 - [`psd.py`](https://github.com/pymanopt/pymanopt/blob/master/pymanopt/manifolds/psd.py) in [pymanopt](https://github.com/pymanopt/pymanopt)
 - [`spd.py`](https://github.com/dalab/matrix-manifolds/blob/master/graphembed/graphembed/manifolds/spd.py) in [matrix-manifolds](https://github.com/dalab/matrix-manifolds)
 - [`psd.py`](https://github.com/Lezcano/geotorch/blob/master/geotorch/psd.py) in [geotorch](https://github.com/Lezcano/geotorch)
 - [`spd_matrices.py`](https://github.com/geomstats/geomstats/blob/master/geomstats/geometry/spd_matrices.py) in [geotorch](https://github.com/geomstats/geomstats) 

### Some paper using SPD Manifolds 
- Computationally Tractable Riemannian Manifolds for Graph Embeddings [[arxiv](https://arxiv.org/abs/2002.08665)]
- A Riemannian Network for SPD Matrix Learning [[arxiv](https://arxiv.org/abs/1608.04233)] 

All suggestions will be accepted with an open mind. Hoping this pull request will be merged when all works have done.

### CODE REVIEW

**Comment by ferrine** [comment]
Hi, sure, I'll review on weekends!

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+        Returns
+        -------
+        torch.Tensor
+            symmetric matrix with function applied to
+        """
+        e, v = torch.symeig(x, eigenvectors=True)
+        return v @ torch.diag_embed(func(e)) @ v.transpose(-1, -2)
+
+    def _expm(self, x: torch.Tensor, using_native=False) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:68]: what is the case when we have to choose? btw, I have an implemented [`expm`](https://github.com/geoopt/geoopt/blob/master/geoopt/linalg/batch_linalg.py#L50). This implementation was waiting for https://github.com/pytorch/pytorch/issues/9983 and seems like it is partially resolved in the new release. Feel free to conditionally replace it (right in `geoopt/linalg/batch_linalg.py`) based on availability of `torch.matrix_exp`.

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        -------
+        torch.Tensor
+            exp(x)
+        """
+        if using_native:
+            return torch.matrix_exp(x)
+        else:
+            return self._funcm(x, torch.exp)
+
+    def _logm(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:88]: Yes, matrix logarithm is not implemented in PyTorch. Personally, I think the best place to put functions like `_funcm`, `_logm` into `geoopt/linalg/batch_linalg.py` and add a comment that this is a temporary workaround awaiting PyTorch upstream

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            log(x)
+        """
+        return self._funcm(x, torch.log)
+
+    def _sqrtm(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:103]: same comments apply here

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            sqrt(x)
+        """
+        return self._funcm(x, torch.sqrt)
+
+    def _invm(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:118]: What is the benchmark against naive matrix inverse? Is it considerably faster or more numerically stable?

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1}
+        """
+        return self._funcm(x, torch.reciprocal)
+
+    def _inv_sqrtm1(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:133]: Well, I believe that is faster since it fuses 2 ops. Am I correct?

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1/2}
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:144]:
```suggestion
            :math:`x^{-1/2}`
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1}
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:129]:
```suggestion
            :math:`x^{-1}`
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            sqrt(x)
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:114]:
```suggestion
            :math:`x^{1/2}`
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            log(x)
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:99]:
```suggestion
            :math:`\log(x)`
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+        torch.Tensor
+            exp(x)
+        """
+        if using_native:
+            return torch.matrix_exp(x)
+        else:
+            return self._funcm(x, torch.exp)
+
+    def _logm(self, x: torch.Tensor) -> torch.Tensor:
+        """Symmetric matrix logarithm.
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:89]:
```suggestion
        r"""Symmetric matrix logarithm.
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+        using_native : bool, optional
+            if using native matrix exponent `torch.matrix_exp`, by default False
+
+        Returns
+        -------
+        torch.Tensor
+            exp(x)
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:81]:
```suggestion
            :math:`\exp(x)`
```

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            x^{-1/2}, sqrt(x)
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:159]:
```suggestion
            :math:`x^{-1/2}`, :math:`x^{1/2}`
```

File: geoopt/manifolds/symmetric_positive_definite.py (hunk 5 above)
Comment (ferrine) [reply geoopt/manifolds/symmetric_positive_definite.py:133]: Also, the best place to put this is batch_linalg file

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+        """
+        e, v = torch.symeig(x, eigenvectors=True)
+        sqrt_e = torch.sqrt(e)
+        inv_sqrt_e = 1 / sqrt_e
+        return (
+            v @ torch.diag_embed(inv_sqrt_e) @ v.transpose(-1, -2),
+            v @ torch.diag_embed(sqrt_e) @ v.transpose(-1, -2),
+        )
+
+    def _sym(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:169]: already [there](https://github.com/geoopt/geoopt/blob/master/geoopt/linalg/batch_linalg.py#L9)

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+        ----------
+        A new metric on the manifold of kernel matrices with application to matrix geometric means. 2012.
+        """.format(
+            self._dist_doc
+        )
+
+        def log_det(tensor: torch.Tensor) -> torch.Tensor:
+            return torch.log(torch.det(tensor))
+
+        if keepdim:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:239]: keep_dim is an important parameter, let's just expand collapsed dims in this case?

File: geoopt/manifolds/symmetric_positive_definite.py [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+            inner product (broadcasted)
+
+        Raises
+        ------
+        ValueError
+            if `keepdim` sine `torch.trace` doesn't support keepdim
+        """
+        if v is None:
+            v = u
+        if keepdim:
```
Comment (ferrine) [inline geoopt/manifolds/symmetric_positive_definite.py:367]: again, this is an important feature used in optimizer routine

**Review by ferrine — CHANGES_REQUESTED** [review]
Hi! This is an amazing PR! It is well written and self-contained. Tests are still a concern, there are none. I see future work as follows:

- [ ] Some functions require to be refactored and moved into the `batch_linalg` module. This will improve reusability and modularity. 
- [ ] You'll need tests to make sure that your implementation is working. I'll write an instruction about this at the end of the review comment.
- [ ] There is missing functionality for inner product and similar methods: `keepdims`. to perform proper broadcasting we'll need this.
- [ ] There is no implementation for `random` and `origin`. It is problematic to create initial points without these functions.

To write a test for a new manifold, you will need:
- [ ] go to https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py#L23 and create a new shape case for a symmetric pd matrix.

This is not yet fully automated, you'll have to create
- [ ] a test case similar to https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py#L40. It should manually compute 
    1. initial point https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py#L43
    2. projection out of this point https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py#L46
    3. same for the tangent vector https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py#L47
 
- [ ] To make sure it is at least numerically stable you can create a simple optimization problem as done for Stiefel Manifold https://github.com/geoopt/geoopt/blob/master/tests/test_rsgd.py#L16

And great! I'm available to chat at almost any time.

File: geoopt/manifolds/symmetric_positive_definite.py (hunk 5 above)
Comment (tao-harald) [reply geoopt/manifolds/symmetric_positive_definite.py:133]: > Well, I believe that is faster since it fuses 2 ops. Am I correct?

Yeah, since some operations need both inverse and inverse square root, it's better to fuse these two operations since they both require eigendecomposition.

**Comment by ferrine** [comment]
Also, I think we can do the same trick as with Stiefel manifold when choosing metric

**Comment by ferrine** [comment]
hi, @tao-harald, what is the status of the PR, is it ready for review and, hopefully, merge?

**Comment by tao-harald** [comment]
> hi, @tao-harald, what is the status of the PR, is it ready for review and, hopefully, merge?

Being busy with study, will back to finish this pull request after the final exam, about Jan. 25th.

**Comment by ferrine** [comment]
That's fine, no rush

**Comment by ferrine** [comment]
Hi, @tao-harald, HOw are you? Hope you are doing well after weekends and exams :slightly_smiling_face: 

Anything I can help with continuing the PR?

**Comment by tao-harald** [comment]
> Hi, @tao-harald, How are you? Hope you are doing well after weekends and exams slightly_smiling_face
> 
> Anything I can help with continuing the PR?

I'm working on this pr now, while there are still problem I cannot solve: This implementation cannot pass the `reversible` test, so at last commit I turn it to `false`. I think that may be caused by a wrong implementation way and I need to read more paper and have some try.

**Comment by fedelopez77** [comment]
Hello!
Thank you so so so much for this implementation!
I am using it at the moment, and so far it seems to be working pretty well.

Just a minor suggestion, that I don't know if it helps, but I implemented "random" by adding small symmetric perturbations to the identity.

**Comment by tao-harald** [comment]
image
That's a milestone. This PR shall be able to merge soon.

**Comment by ferrine** [comment]
:tada:
